### DISCUSSION TITLE
Use timestamps on Map Matching

### DISCUSSION THREAD

**Discussion by mushabtinumbang** [valhalla/valhalla#4731] [Q&A] [answered]
Hi, i am wondering if we can use timestamps informations for each of the matched points. I am currently using `trace_attributes` to get detailed informations about the edges, points, etc. I have tried to set use_timestamps = True but can't see any difference from the output.

```
"shape_match": "map_snap",
"trace_options": {
              "gps_accuracy": int(temp_gps_accuracy), 
              "search_radius": int(search_radius),
              "breakage_distance": int(breakage_distance),
              "interpolation_distance": int(interpolation_distance),
          },
"costing": f"{vehicle_type_arg}",
"filters": {"attributes": [
            "matched.point", 
            "matched.edge_index", 
            "edge.toll", 
            "edge.vehicle_type", 
            "edge.id",
            "edge.speed",
            "admin.state_text",
            "matched.type",
            "edge.way_id",
            "shape"
        ], "action": "include"},
"shape": [{
            "lat": float(temp_lat[i]),  
            "lon": float(temp_lon[i]),
            "time": int(temp_time[i]), 
            "type": "break" if (i == 0) or (i == len(temp_points) - 1) else "via" 
    
          } for i in range(len(temp_points))],
"begin_time": int(temp_time[0]), 
"use_timestamps": True,
"directions_type": "none"
```

For more context on the problem, i have this timestamp information for each points in my original data. I wanted to atleast have specific timestamps for each points in the map matching result just like my original dataset to not lose any context / informations. Is there anything that I can do to solve this problem? Would appreciate if anyone can help on this one 👍 , thanks a lot

**Comment by chrstnbwnkl** [comment, marked as answer]
Looks like the attribute you're looking for is `node.elapsed_time`, which should give you something like this on each edge object in the output: 

```
"end_node": {
        "elapsed_time": 345,
        "elapsed_cost": 448.205
}
```

**Comment by bmionescu** [comment]
I want to add to this. In Valhalla's response, the 'shape' key gives you a polyline which can be decoded to get a list of map-matched coordinates. And, the 'edges' key gives you an object which has an 'end_node' key, which leads to 'elapsed_time' as mentioned above. However, the number of elapsed_times does not necessarily match the number of coordinates in the decoded polyline. If the number of coordinates does not match the number of timestamps, it's impossible to piece together what timestamp goes with what coordinate. But, what I found is that you can ignore the 'shape' polyline, and instead use the 'matched_points' key. The length of matched_points matches the number of raw data points you inputted to the request to Valhalla, BUT, each matched_point comes with an 'edge_index' property, which does actually correspond to the number of timestamps.
